native/service/cross_chain_manager/ripple/ripple_handler.go

```diff
@@ -0,0 +1,285 @@
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"math/big"
+
+	"github.com/polynetwork/poly/common"
+	"github.com/polynetwork/poly/native"
+	"github.com/polynetwork/poly/native/event"
+	crosscommon "github.com/polynetwork/poly/native/service/cross_chain_manager/common"
+	"github.com/polynetwork/poly/native/service/governance/side_chain_manager"
+	"github.com/polynetwork/poly/native/service/utils"
+	"github.com/polynetwork/ripple-sdk/types"
+	"github.com/rubblelabs/ripple/data"
+)
+
+type RippleHandler struct {
+}
+
+func NewRippleHandler() *RippleHandler {
+	return &RippleHandler{}
+}
+
+func (this *RippleHandler) MultiSign(service *native.NativeService) error {
+	params := new(MultiSignParam)
+	if err := params.Deserialization(common.NewZeroCopySource(service.GetInput())); err != nil {
+		return fmt.Errorf("MultiSign, contract params deserialize error: %v", err)
+	}
+
+	// get rippleExtraInfo
+	rippleExtraInfo, err := side_chain_manager.GetRippleExtraInfo(service, params.ToChainId, params.AssetAddress)
+	if err != nil {
+		return fmt.Errorf("MultiSign, side_chain_manager.GetRippleExtraInfo error")
+	}
+
+	// get raw txJsonInfo
+	raw, err := GetTxJsonInfo(service, params.FromChainId, params.TxHash)
+	if err != nil {
+		return fmt.Errorf("MultiSign, get txJsonInfo error")
+	}
+
+	// check if aleady done
+	multisignInfo, err := GetMultisignInfo(service, raw)
+	if err != nil {
+		return fmt.Errorf("MultiSign, GetMultisignInfo error")
+	}
+	if multisignInfo.Status {
+		return nil
+	}
+
+	// check if signature is valid
+	txJson := new(types.MultisignPayment)
+	err = json.Unmarshal([]byte(params.TxJson), txJson)
+	if err != nil {
+		return fmt.Errorf("MultiSign, unmarshal signed txjson error: %s", err)
+	}
+	for _, s := range txJson.Signers {
+		signerAccount, err := data.NewAccountFromAddress(s.Signer.Account)
+		if err != nil {
+			return fmt.Errorf("MultiSign, data.NewAccountFromAddress error: %s", err)
+		}
+		signerPk, err := hex.DecodeString(s.Signer.SigningPubKey)
+		if err != nil {
+			return fmt.Errorf("MultiSign, hex.DecodeString signer pk error: %s", err)
+		}
+		signature, err := hex.DecodeString(s.Signer.TxnSignature)
+		if err != nil {
+			return fmt.Errorf("MultiSign, hex.DecodeString signature error: %s", err)
+		}
+
+		// check if valid signer
+		flag := false
+		for _, v := range rippleExtraInfo.Pks {
+			if fmt.Sprintf("%X", v) == s.Signer.SigningPubKey {
+				flag = true
+			}
+		}
+		if !flag {
+			return fmt.Errorf("MultiSign, signer is not multisign account")
+		}
+
+		//check if valid signature
+		err = types.CheckMultiSign(raw, *signerAccount, signerPk, signature)
+		if err != nil {
+			return fmt.Errorf("MultiSign, types.CheckMultiSign error: %s", err)
+		}
+		signer := &Signer{
+			Account:       signerAccount.Bytes(),
+			TxnSignature:  signature,
+			SigningPubKey: signerPk,
+		}
+		sink := common.NewZeroCopySink(nil)
+		signer.Serialization(sink)
+		multisignInfo.SigMap[hex.EncodeToString(sink.Bytes())] = true
```

Comment:
Possible multiple signatures with same account?

Reply:
If multiple signatures with same account and same raw msg, the signature after signed is the same too, so the map will be covered and length count do not +1